model/riscv_insts_vext_arith.sail

```diff
@@ -0,0 +1,496 @@
+    if mask[i] then {
+      result[i] = match funct6 {
+        VI_VSLIDEUP    => {
+                            assert(vs2 != vd);
+                            if i >= imm_val then vs2_val[i - imm_val] else vd_val[i]
+                          },
+        VI_VSLIDEDOWN  => {
+                            let VLMAX = int_power(2, LMUL_pow + VLEN_pow - SEW_pow);
+                            assert(VLMAX > 0 & VLMAX <= 'n);
+                            if i + imm_val < VLMAX then vs2_val[i + imm_val] else zeros()
+                          },
+        VI_VRGATHER    => {
+                            assert(vs2 != vd);
+                            let VLMAX = int_power(2, LMUL_pow + VLEN_pow - SEW_pow);
+                            assert(VLMAX > 0 & VLMAX <= 'n);
+                            if imm_val < VLMAX then vs2_val[imm_val] else zeros() 
+                          }
+      }
+    }
+  };
+
+  write_vreg(num_elem, SEW, LMUL_pow, vd, result);
+  vstart = EXTZ(0b0);
+  RETIRE_SUCCESS
+} 
+
+mapping visg_mnemonic : visgfunct6 <-> string = {
+  VI_VSLIDEUP     <-> "vslideup.vi",
+  VI_VSLIDEDOWN   <-> "vslidedown.vi",
+  VI_VRGATHER     <-> "vrgather.vi"
+}
+
+mapping clause assembly = VISG(funct6, vm, vs2, simm, vd)
+  <-> visg_mnemonic(funct6) ^ spc() ^ vreg_name(vd) ^ sep() ^ vreg_name(vs2) ^ sep() ^ reg_name(simm) ^ maybe_vmask(vm)
+
+/* ********************* Whole Vector Register Move (OPIVI) ********************** */
+union clause ast = VMVRTYPE : (regidx, bits(5), regidx)
+
+mapping clause encdec = VMVRTYPE(vs2, simm, vd)
+  <-> 0b100111 @ 0b1 @ vs2 @ simm @ 0b011 @ vd @ 0b1010111
+
+function clause execute(VMVRTYPE(vs2, simm, vd)) = {
+  let SEW     = get_sew();
+  let imm_val = unsigned(EXTZ(sizeof(xlen), simm));
+  let EMUL    = imm_val + 1;
+
+  assert(EMUL == 1 | EMUL == 2 | EMUL == 4 | EMUL == 8);
+  let EMUL_pow = log2(EMUL);
+  let num_elem = get_num_elem(EMUL_pow, SEW);
+  let 'n = num_elem;
+  let 'm = SEW;
+
+  let vs2_val : vector('n, dec, bits('m)) = read_vreg(num_elem, SEW, EMUL_pow, vs2);
+  write_vreg(num_elem, SEW, EMUL_pow, vd, vs2_val);
+  vstart = EXTZ(0b0);
+
+  RETIRE_SUCCESS
+} 
+
+mapping simm_string : bits(5) <-> string = {
+  0b00000 <-> "1",
+  0b00001 <-> "2",
+  0b00011 <-> "4",
+  0b00111 <-> "8"
+}
+
+mapping clause assembly = VMVRTYPE(vs2, simm, vd)
+  <-> "vmv" ^ simm_string(simm) ^ "r.v" ^ spc() ^ vreg_name(vd) ^ sep() ^ vreg_name(vs2)
+
+/* ****************************** OPMVV (VWXUNARY0) ****************************** */
+union clause ast = VMVXS : (regidx, regidx)
+
+mapping clause encdec = VMVXS(vs2, rd)
+  <-> 0b010000 @ 0b1 @ vs2 @ 0b00000 @ 0b010 @ rd @ 0b1010111
+
+function clause execute(VMVXS(vs2, rd)) = {
+  let SEW      = get_sew();
+  let num_elem = get_num_elem(0, SEW);
+
+  assert(num_elem > 0);
+  let 'n = num_elem;
+  let 'm = SEW;
+
+  let vs2_val : vector('n, dec, bits('m)) = read_vreg(num_elem, SEW, 0, vs2);
+  if sizeof(xlen) < SEW then X(rd) = slice(vs2_val[0], 0, sizeof(xlen))
```

Comment:
Generally clearer to not have X(rd) being assigned to inside an if/else unless the assignment is conditional. Otherwise it's confusing and you have to stop and think about whether it's always assigned to or not. Can either do
```
let result = if ...
X(rd) = result
```
or
```
X(rd) = if ...
```

Reply:
This assignment is modified.